[inductor] Add type hints to functions in decompositions.py

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131780

Summary: ATT

Test Plan: lintrunner

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

aakhundov committed:
[inductor] Add type hints to functions in decompositions.py

Summary: ATT

Test Plan: lintrunner

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

aakhundov committed:
Update on "[inductor] Add type hints to functions in decompositions.py"


Summary: ATT

Test Plan: lintrunner

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -109,8 +108,8 @@
 
-def register_decomposition(ops):
-    for op in [ops] if callable(ops) else ops:
+def register_decomposition(ops: List[torch._ops.OperatorBase]) -> Callable[..., Any]:
```
This should be `Union[OpOverloadPacket, OperatorBase]`

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -119,24 +118,33 @@ def register_decomposition(ops):
-def clamp(x, min=None, max=None):
+def clamp(
+    x: torch.Tensor,
+    min: Optional[Union[int, float]] = None,
+    max: Optional[Union[int, float]] = None,
```
Can we use `torch.types.Number` here ?

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -145,7 +153,11 @@ def clamp(x, min=None, max=None):
-def full(size, fill_value, **kwargs):
+def full(
+    size: List[Union[int, torch.SymInt]],
+    fill_value: Union[int, float],
```
ditto

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -282,10 +310,13 @@ def mm(self, input2):
 @register_decomposition([aten.cat.default])
-def cat(tensors, dim=0):
+def cat(
+    tensors: torch.Tensor,
```
this should be list of tensors right ?

aakhundov replied:
Yep, great catch!

aakhundov replied:
Wondering how did mypy not complain...

eellison replied:
We probably dont invoke this function directly much.. only through decomps

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -342,7 +373,12 @@ def angle(x):
+    x: torch.Tensor,
+    y: torch.Tensor,
+    *,
+    alpha: Optional[float] = None,
```
alpha `float` -> Number

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
 @register_decomposition(aten._foreach_addcmul.Scalar)
-def _foreach_addcmul_scalar(self, left_tensors, right_tensors, scalar=1):
+def _foreach_addcmul_scalar(
+    self: torch.Tensor,
```
self is a `List[Tensor]`

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
 @register_decomposition(aten._foreach_addcdiv.Scalar)
-def _foreach_addcdiv_scalar(self, left_tensors, right_tensors, scalar=1):
+def _foreach_addcdiv_scalar(
+    self: torch.Tensor,
```
ditto

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
+def _foreach_lerp_scalar(
+    start_tensors: List[torch.Tensor],
+    end_tensors: List[torch.Tensor],
+    weight: float,
```
weight -> Number

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -634,19 +748,23 @@ def miopen_batch_norm(
 
 @functools.lru_cache(None)
-def fast_random_decomps():
+def fast_random_decomps():  # type: ignore[no-untyped-def]
```
`-> Dict[OpOverload, Callable[[Any], Any]` ?

aakhundov replied:
So I've tried this and a few external call sites have started breaking. Let me try again.

aakhundov replied:
So the best I could do is `Dict[Any, Callable[..., Any]]` without ripples in other files. Left a TODO.

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -634,19 +748,23 @@ def miopen_batch_norm(
 
 
-def select_decomp_table():
+def select_decomp_table():  # type: ignore[no-untyped-def]
```
same

eellison reviewed (approved):
looks good! a few comments

eellison reviewed (approved):
looks good! a few comments

aakhundov committed:
Update on "[inductor] Add type hints to functions in decompositions.py"


Summary: ATT

Test Plan: lintrunner

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
